datalad/local/credentials.py

```diff
@@ -0,0 +1,519 @@
+            self._unset_credprops(name, nonglobal_remove_props, 'local')
+        nonremoved_props = self._props_defined_in_cfg(
+            name, nonglobal_remove_props)
+        if nonremoved_props:
+            raise RuntimeError(
+                f"Cannot remove properties {nonglobal_remove_props} "
+                f"for credential {name!r}, defined outside global or local "
+                "configuration scope. Remove manually")
+        updated = {k: None for k in remove_props}
+
+        # set non-secret props
+        #
+        set_props = {
+            k: v for k, v in cred.items()
+            if v is not None and k != 'secret'
+        }
+        for k, v in set_props.items():
+            var = _get_cred_cfg_var(name, k)
+            if self._cfg.get(var) == v:
+                # desired value already exists, we are not
+                # storing again to preserve the scope it
+                # was defined in
+                continue
+            # we always write to the global scope (ie. user config)
+            # credentials are typically a personal, not a repository
+            # specific entity -- likewise secrets go into a personal
+            # not repository-specific store
+            # for custom needs users can directly set the respective
+            # config
+            self._cfg.set(var, v, scope='global', force=True, reload=False)
+            updated[k] = v
+        if set_props:
+            self._cfg.reload()
+
+        # set secret
+        #
+        # we aim to update the secret in the store, hence we must
+        # query for a previous setting in order to reliably report
+        # updates
+        prev_secret = self._get_secret_from_keyring(name, type_hint)
+        if 'secret' not in cred:
+            # we have no removal directive, reuse previous secret
+            cred['secret'] = prev_secret
+        if cred.get('secret') is None:
+            # we want to reset the secret, consider active config
+            cred['secret'] = \
+                self._cfg.get(_get_cred_cfg_var(name, 'secret'))
+
+        if cred.get('secret') is None:
+            # we have no secret specified or in the store already: ask
+            # typically we would end up here with an explicit attempt
+            # to set a credential in a context that is known to an
+            # interactive user, hence the messaging here can be simple
+            cred['secret'] = ui.question(
+                'secret',
+                title=f"Please enter secret for credential {name!r}",
+                repeat=self._cfg.obtain(
+                    'datalad.credentials.repeat-secret-entry'),
+                hidden=self._cfg.obtain(
+                    'datalad.credentials.hidden-secret-entry'),
+            )
+        # at this point we will have a secret. it could be from ENV
+        # or provided, or entered. we always want to put it in the
+        # store
+        self._keyring.set(name, 'secret', cred['secret'])
```

Comment:
Since the entire thing seems to aim at migrating existing credentials, I think we may want to also try and get rid of using `datalad.support.keyring_._get_service_name` (which `set` does use). I don't see any value in a mandatory `datalad-` prefix, ruling out to use credentials from an existing keychain, just b/c they lack this prefix. I think we'd want to be able to point to any existing one that is accessible to us.